torch/csrc/jit/fuser/kernel_cache.cpp

```diff
@@ -0,0 +1,36 @@
+#include "torch/csrc/jit/fuser/kernel_cache.h"
+
+#include <unordered_map>
+#include <mutex>
+#include <cstdint>
+
+namespace torch { namespace jit { namespace fuser {
+
+// Note: std::unordered_map does not invalidate references even if rehashing
+// occurs. This is a critical property for thread-safety. 
+static int64_t kernel_counter{0};
+static std::unordered_map<int64_t, KernelSpec> specMap_;
+static std::mutex mutex_;
```

Comment:
It would be better to put those in a struct, and then define it as a `static` variable inside a function that returns it. That would saves us from potential static initialization order issues in the future.

Reply:
Will do.